File: libs/server/api/src/api/model/edge_model.rs
use serde::Serialize;
use std::collections::BTreeMap;

use crate::domain::{DiagramEdge, JsonObject, Ref};

use super::super::links::{
    workspace_diagram_edge_href, workspace_diagram_edges_href, workspace_diagram_href, Link,
};

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(in crate::api) struct EdgeModel {
    #[serde(rename = "_links")]
    links: BTreeMap<String, Link>,
    id: String,
    source: Ref<String>,
    target: Ref<String>,
    logical_relationship: Option<Ref<String>>,
    source_handle: Option<String>,
    target_handle: Option<String>,
    kind: Option<String>,
    style: JsonObject,
    data: JsonObject,
    animated: bool,
    hidden: bool,
    marker_start: Option<JsonObject>,
    marker_end: Option<JsonObject>,
    path_options: JsonObject,
    interaction_width: Option<f64>,
    created_at: String,
    updated_at: String,
}

pub(in crate::api) fn edge_model(workspace_id: &str, edge: &DiagramEdge) -> EdgeModel {
    let diagram_id = edge.diagram_id();
    let edge_id = edge.identity();
    let description = edge.description();

    EdgeModel {
        links: BTreeMap::from([
            (
                "self".to_string(),
                Link::new(workspace_diagram_edge_href(
                    workspace_id,
                    diagram_id,
                    edge_id,
                )),
            ),
            (
                "collection".to_string(),
                Link::new(workspace_diagram_edges_href(workspace_id, diagram_id)),
            ),
            (
                "diagram".to_string(),
                Link::new(workspace_diagram_href(workspace_id, diagram_id)),
            ),
        ]),
        id: edge_id.to_string(),
        source: description.source.clone(),
        target: description.target.clone(),
        logical_relationship: description.logical_relationship.clone(),
        source_handle: description.source_handle.clone(),
        target_handle: description.target_handle.clone(),
        kind: description.kind.clone(),
        style: description.style.clone(),
        data: description.data.clone(),
        animated: description.animated,
        hidden: description.hidden,
        marker_start: description.marker_start.clone(),
        marker_end: description.marker_end.clone(),
        path_options: description.path_options.clone(),
        interaction_width: description.interaction_width,
        created_at: edge.created_at().to_string(),
        updated_at: edge.updated_at().to_string(),
    }
}

File: libs/server/api/src/api/model/edge-model.ts
import { DiagramEdge, JsonObject, Ref } from '@evidence/server-domain';
import {
  link,
  Link,
  workspaceDiagramEdgeHref,
  workspaceDiagramEdgesHref,
  workspaceDiagramHref,
} from '../links';

export interface EdgeModel {
  _links: Record<string, Link>;
  id: string;
  source: Ref<string>;
  target: Ref<string>;
  logicalRelationship: Ref<string> | null;
  sourceHandle: string | null;
  targetHandle: string | null;
  kind: string | null;
  style: JsonObject;
  data: JsonObject;
  animated: boolean;
  hidden: boolean;
  markerStart: JsonObject | null;
  markerEnd: JsonObject | null;
  pathOptions: JsonObject;
  interactionWidth: number | null;
  createdAt: string;
  updatedAt: string;
}

export function edgeModel(workspaceId: string, edge: DiagramEdge): EdgeModel {
  const edgeId = edge.identity();
  const description = edge.description();
  return {
    _links: {
      self: link(workspaceDiagramEdgeHref(workspaceId, edgeId)),
      collection: link(workspaceDiagramEdgesHref(workspaceId)),
      diagram: link(workspaceDiagramHref(workspaceId)),
    },
    id: edgeId,
    source: description.source,
    target: description.target,
    logicalRelationship: description.logicalRelationship,
    sourceHandle: description.sourceHandle,
    targetHandle: description.targetHandle,
    kind: description.kind,
    style: description.style,
    data: description.data,
    animated: description.animated,
    hidden: description.hidden,
    markerStart: description.markerStart,
    markerEnd: description.markerEnd,
    pathOptions: description.pathOptions,
    interactionWidth: description.interactionWidth,
    createdAt: description.createdAt,
    updatedAt: description.updatedAt,
  };
}
